get_fm_data.py

```diff
@@ -0,0 +1,200 @@
+        if res.status_code == 401:
+            self.handle_response_errors(message, res, "get EPEX prices", self.get_epex_prices, *args, **kwargs)
+            return
+        # Do not count authoristion errors int he attempts.
+        self.attempts_today += 1
+
+        # A bit prematurely calculate a delay in seconds in which to call
+        # this  function again incase something goes wrong.
+        # The delays total to 11h30m with max 12 attempts.
+        delay = (5 + self.attempts_today * 8) * 60
+
+        if res.status_code != 200:
+            self.log(f"Get FM EPEX data failed with status_code: {res.status_code}, full response was {res.json()}. Retry in {delay} minutes.")
+            self.run_in(self.get_epex_prices, delay)
+            return
+
+        prices = res.json()
+
+        # We might get older data, but if the UI has no data yet it is still
+        # nice to show this older data anayway, so process it.
+
+        # Not so usefull checks??
+        # epex_entity = self.get_state("input_text.epex_prices", attribute='records')
+        # if epex_entity != None:
+        #     #There are records already no need to process data
+        #     return
+        # self.log("Eventhough prices are not new, there are currently no prices registered at all (due to restart?), so process data anyway.")
+
+        # From FM format (€/MWh) to user desired format (€ct/kWh) 
+        # = * 100/1000 = 1/10. Also include VAT
+        VAT = float(self.args["VAT_NL"])
+        conversion = 1/10 * VAT
+        # For NL electricity is a markup for sustainability 
+        markup = float(self.args["markup_per_kWh_NL"])
+        epex_price_points = []
+        has_negative_prices = False
+        for price in prices:
+            data_point = {}
+            data_point['time'] = datetime.fromtimestamp(price['event_start']/1000).isoformat()
+            data_point['price'] = round((price['event_value'] * conversion) + markup, 2)
+            if data_point['price'] < 0:
+                has_negative_prices = True
+            epex_price_points.append(data_point)
+
+        # To make sure HA considers this as new info a datetime is added
+        new_state = "APEX prices collected at " + now.isoformat()
+        result = {}
+        result['records'] = epex_price_points
+        self.set_state("input_text.epex_prices", state=new_state, attributes=result)
+
+        # FM returns all the prices it has, sometimes it has not retreived new
+        # prices yet, than it communicates the prices it does have.
+        date_latest_price = datetime.fromtimestamp(prices[-1].get('event_start')/1000).isoformat()
+        date_tomorrow = (now + timedelta(days=1)).isoformat()
+        if date_latest_price < date_tomorrow:
+            self.log(f"FM Epex prices seem not renewed yet, latest price at: {date_latest_price}, retry in {delay} minutes.")
+            self.run_in(self.get_epex_prices, delay)
+        else:
+            if has_negative_prices == True:
```

Comment:
Can be simplified to `if has_negative_prices:`

Reply:
Done..